File: packages/sdk-js/src/agent.ts
import { AAPAgentConfig, AgentIdentity, MessageHandler } from './types';
import { AAPSession } from './session';
import { DEFAULT_REGISTRY_URL } from './constants';
import {
  generateKeyPair,
  sign,
  generateNonce,
  validateTimestamp,
  dilithium3KeyPair,
  dilithium3Sign,
} from '@aap/crypto';

export class AAPAgent {
  private config: AAPAgentConfig;
  private identity?: AgentIdentity;
  private privateKeyHex?: string;
  private messageHandler?: MessageHandler;
  private registryUrl: string;

  constructor(config: AAPAgentConfig) {
    this.config = config;
    this.registryUrl = config.registryUrl || DEFAULT_REGISTRY_URL;
    if (config.onMessage) this.messageHandler = config.onMessage;
  }

  /**
   * Register this agent with the AAP Registry.
   * Generates a real Ed25519 key pair, signs the registration body,
   * and publishes the aap:// address.
   */
  async register(): Promise<AgentIdentity> {
    // Generate key pair — Ed25519 (default) or Dilithium3 (post-quantum)
    const algo = this.config.signatureAlgorithm ?? 'ed25519';
    const kp = algo === 'dilithium3' ? dilithium3KeyPair() : generateKeyPair();
    this.privateKeyHex = kp.privateKeyHex;

    const aap_address  = `aap://${this.config.name}`;
    const endpoint_url = `https://agent.${this.config.name.replace(/\./g, '-')}.local`;
    const timestamp    = Date.now();
    const nonce        = generateNonce();

    // Build body without signature for signing
    const bodyToSign = {
      aap_address,
      public_key_hex:  kp.publicKeyHex,
      endpoint_url,
      capabilities:    this.config.capabilities || [],
      owner_type:      'human',
      timestamp,
      nonce,
    };

    // Sign the body with the chosen algorithm
    const messageBytes = new TextEncoder().encode(JSON.stringify(bodyToSign));
    const signature = algo === 'dilithium3'
      ? dilithium3Sign(messageBytes, kp.privateKeyHex)
      : sign(messageBytes, kp.privateKeyHex);

    const response = await fetch(`${this.registryUrl}/v1/register`, {
      method:  'POST',
      headers: { 'Content-Type': 'application/json' },
      body:    JSON.stringify({ ...bodyToSign, signature }),
    });

    if (!response.ok) {
      const err = await response.json() as { error: string };
      throw new Error(`Registration failed: ${err.error}`);
    }

    const data = await response.json() as { did: string; aap_address: string };
    this.identity = { ...data, public_key_hex: kp.publicKeyHex };

    console.log(`✅ Registered: ${aap_address}`);
    console.log(`   DID: ${this.identity.did}`);
    return this.identity;
  }

  /**
   * Connect to another agent. Looks up their DID document,
   * verifies their identity, and performs the AAP handshake.
   */
  async connect(address: string): Promise<AAPSession> {
    if (!this.identity || !this.privateKeyHex) {
      throw new Error('Agent not registered. Call register() first.');
    }
    const session = new AAPSession(
      this.identity,
      this.privateKeyHex,
      address,
      this.registryUrl
    );
    await session.handshake();
    return session;
  }

  /** Register a handler for incoming messages. */
  onMessage(handler: MessageHandler): void {
    this.messageHandler = handler;
  }

  /**
   * Returns a default L2 task manifest for the configured capabilities.
   */
  getCapabilityManifest() {
    const caps = this.config.capabilities?.map(c => c.toString()) ?? [];

    // Always include PING + DISCONNECT + ERROR as baseline
    const base = new Set(['PING', 'PONG', 'DISCONNECT', 'ERROR', ...caps]);

    // Auto-include paired response actions
    const pairs: Record<string, string> = {
      REQUEST_DATA:        'RETURN_DATA',
      REQUEST_QUOTE:       'RETURN_QUOTE',
      DELEGATE_TASK:       'TASK_RESULT',
      PROPOSE_TRANSACTION: 'ACCEPT_TRANSACTION',
      REQUEST_HUMAN_AUTH:  'HUMAN_AUTH_RESPONSE',
      HANDSHAKE_INIT:      'HANDSHAKE_RESPONSE',
    };
    for (const [req, res] of Object.entries(pairs)) {
      if (base.has(req)) base.add(res);
    }

    return {
      agent_did:          this.identity?.did ?? `did:aap:${this.config.name}`,
      level:              2,
      allowed_actions:    Array.from(base),
      denied_actions:     [],
      allowed_data_types: [],
      denied_data_types:  [],
      approved_agents:    [],
      expires_at:         new Date(Date.now() + 24 * 60 * 60 * 1000).toISOString(),
    };
  }

  /** Look up any agent by aap:// address. */
  static async lookup(address: string, registryUrl = DEFAULT_REGISTRY_URL) {
    const encoded  = encodeURIComponent(address);
    const response = await fetch(`${registryUrl}/v1/lookup/${encoded}`);
    if (!response.ok) throw new Error(`Agent not found: ${address}`);
    return response.json();
  }

  /** Verify another agent's identity and trust score. */
  async verify(did: string) {
    const response = await fetch(`${this.registryUrl}/v1/agent/${did}/trust`);
    if (!response.ok) throw new Error(`Could not verify DID: ${did}`);
    return response.json();
  }

  getIdentity(): AgentIdentity | undefined { return this.identity; }
}

File: packages/sdk-python/src/aap/agent.py
import secrets
import httpx
from typing import Optional, Callable, Awaitable
from .constants import DEFAULT_REGISTRY_URL
from .session import AAPSession


class AAPAgent:
    """
    AAP Agent — the main entry point for the Python SDK.

    Usage:
        agent = AAPAgent(name='vimalesh.finance', capabilities=[Capability.REQUEST_QUOTE])
        await agent.register()
        session = await agent.connect('aap://demo.echo')
        result  = await session.send(Action.PING, {})
    """

    def __init__(
        self,
        name: str,
        capabilities: Optional[list] = None,
        on_message: Optional[Callable] = None,
        registry_url: str = DEFAULT_REGISTRY_URL,
    ):
        self.name = name
        self.capabilities = capabilities or []
        self.on_message = on_message
        self.registry_url = registry_url
        self.identity: Optional[dict] = None
        self._public_key_hex: Optional[str] = None
        self._private_key_hex: Optional[str] = None

    async def register(self) -> dict:
        """
        Register this agent with the AAP Registry.
        Generates a key pair and publishes the aap:// address.
        """
        self._public_key_hex, self._private_key_hex = self._generate_key_pair()
        aap_address = f"aap://{self.name}"
        endpoint_url = f"https://agent.{self.name.replace('.', '-')}.local"

        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{self.registry_url}/v1/register",
                json={
                    "aap_address": aap_address,
                    "public_key_hex": self._public_key_hex,
                    "endpoint_url": endpoint_url,
                    "capabilities": [str(c) for c in self.capabilities],
                    "owner_type": "human",
                    "signature": "placeholder",  # TODO: Dilithium3 signature
                },
            )
            response.raise_for_status()
            data = response.json()

        self.identity = {**data, "public_key_hex": self._public_key_hex}
        print(f"✅ Registered: {aap_address}")
        print(f"   DID: {self.identity['did']}")
        return self.identity

    async def connect(self, address: str) -> "AAPSession":
        """Connect to another agent. Performs the AAP handshake."""
        if not self.identity:
            raise RuntimeError("Agent not registered. Call register() first.")
        session = AAPSession(self.identity, address, self.registry_url)
        await session.handshake()
        return session

    async def verify(self, did: str) -> dict:
        """Get trust score and verification level for any agent DID."""
        async with httpx.AsyncClient() as client:
            response = await client.get(f"{self.registry_url}/v1/agent/{did}/trust")
            response.raise_for_status()
            return response.json()

    @staticmethod
    async def lookup(address: str, registry_url: str = DEFAULT_REGISTRY_URL) -> dict:
        """Look up any agent by aap:// address."""
        import urllib.parse
        encoded = urllib.parse.quote(address, safe="")
        async with httpx.AsyncClient() as client:
            response = await client.get(f"{registry_url}/v1/lookup/{encoded}")
            response.raise_for_status()
            return response.json()

    def get_identity(self) -> Optional[dict]:
        return self.identity

    def _generate_key_pair(self) -> tuple[str, str]:
        # Placeholder. Production: use pyca/cryptography Dilithium3 or Ed25519.
        pub = "ed25519_pub_" + secrets.token_hex(32)
        prv = "ed25519_prv_" + secrets.token_hex(32)
        return pub, prv
